tests/test_class_sh_unique_ptr_custom_deleter.py

```diff
@@ -0,0 +1,6 @@
+from pybind11_tests import class_sh_unique_ptr_custom_deleter as m
+
+
+def test_create():
+    pet = m.create("abc")
+    assert pet.name == "abc"
```

Comment:
This unit test doesn't actually test the deleter is ran, unless ran inside valgrind / lsan, is that intentional?

Reply:
> This unit test doesn't actually test the deleter is ran, unless ran inside valgrind / lsan, is that intentional?

@wangxf123456 was completely focused on reducing a test from the wild, just enough to reproduce the error message we saw.

Copy-pasting what @iwanders wrote in another comment:

> my main concern is the above statement that the deleter is only handled when going into the smart holder, not when being converted back to a unique pointer.

I agree. I was also completely focused on the issue in the wild. — Reporting a thought/uncertainty in the back of my mind: Do we actually have to solve this (is that feature actually needed in the wild)? Or would it be sufficient to clearly report an error if someone tried to use the non-existing feature? And then work on it as needed.

I'll add a note about it to the PR description, then merge this PR when I see that the GHA are happy.

> Sure, I can't commit a lot of time, but happy to help over the next week or so.

I really appreciate any help! And I'm a big believer in small steps. I want to encourage you to pick out one thing at a time. Solve it. Merge. Next one. (My goal, in general, is to review PRs within hours; highest priority.)

Reply:
> Or would it be sufficient to clearly report an error if someone tried to use the non-existing feature? And then work on it as needed.

Yeah, tricky, it's a balance. I'd at least throw in case there's a non default deleter detected in `as_unique_ptr`, that way it explicitly fails, instead of implicitly losing the deleter and leaking memory, and that should still be doable while avoiding most of the extra code I had to add to account for it.

Speaking of `as_unique_ptr`, another random thought; that method could be marked `&&`, so `std::unique_ptr<T, D> as_unique_ptr() &&`, that ensures it can only be called on an r-value `smart_holder`, which makes it clear that the `smart_holder` object after the method call is in an indeterminate state. In that case it would be illegal to call `as_unique_ptr` on an lvalue though, so it makes it a bit more cumbersome at the callsite, but it does draw more attention to what it is doing. Not saying this should be done, but just stating it is an option to make it clear we are moving out of the smart holder.

> I'll add a note about it to the PR description, then merge this PR when I see that the GHA are happy.

Sounds good.

Reply:
Ah I see that `as_unique_ptr` was also wrapped with `PYBIND11_TESTS_PURE_CPP_SMART_HOLDER_POC_TEST_CPP`; 100% agree it's probably not worth worrying about the deleter there then :+1: